cypress-io/cypress
# misc: Update supported browser messages

- Closes [https://github.com/cypress-io/cypress/issues/28914](https://github.com/cypress-io/cypress/issues/28914)

### Additional details

Current messages for browser support/browser's not found can be confusing e.g.:
`Browser: chrome was not found on your system or is not supported by Cypress.`
This PR aims to provide clear messaging if a browser is not supported, or if it was not found by name or path.
- If a user enters a browser name or path that does not contain any supported browser names, or does not contain any custom browser (e.g. Brave) names added via Cypress config, then the user will see a `BROWSER_NOT_SUPPORTED` error message.
- I have added a regex that should cast a wide net for supported browsers - we ideally want to have zero false negatives. If a user enters a browser name or path that is supported (e.g. `C:\some\path\msedge.exe`) but is not found on their system, they will still see the appropriate `BROWSER_NOT_FOUND_BY_NAME` or `BROWSER_NOT_FOUND_BY_PATH` error. However, if a user misspells a browser name e.g. `webkitt`, or enters an unsupported browser name e.g. `arc`, they will see the `BROWSER_NOT_SUPPORTED` error message.
- I have moved the special `canary` error message to the `BROWSER_NOT_SUPPORTED` error message.
- `getBrowserLauncher()` now throws the `BROWSER_NOT_SUPPORTED` error instead of `BROWSER_NOT_FOUND_BY_NAME` - if `electron` or one of the supported browser families is not found, then it should be safe to assume that this browser is not supported.
- I have added `webkit` to the list of supported browsers in the `BROWSER_NOT_SUPPORTED` error message.
- Custom browsers added via Cypress config still work and are not filtered out.
- I have updated the `BROWSER_NOT_FOUND_BY_NAME` error - this error message may still need to be made clearer.
- This PR doesn't take channels into account e.g. passing in `chrome:someunsupportedchannel` - this will throw a `BROWSER_NOT_FOUND_BY_NAME` error. I think this behavior makes sense as we want to avoid any false negatives - if the name/path contains e.g. `chrome`, we can assume it's supported and then throw a more appropriate error later on.
- _*(Apologies for the number of commits)*_

### Steps to test

Test the following commands:
- `yarn cypress:run --browser unsupportedbrowser --project /your/project`
- `yarn cypress:run --browser badname:channel--project /your/project`
- `yarn cypress:run --browser /path/to/unsupported/browser --project /your/project`
- Run `yarn test test/unit/browsers/browsers_spec.js` to verify passing tests
- Run `yarn test` in `packages/errors` to verify passing tests

### How has the user experience changed?

- If a user enters a browser name or path that does not match any supported browser names, or does not match any custom browsers (e.g. Brave) added in via Cypress config, then the user will see a 'browser not supported' error message.
<img width="608" alt="unsupported_path" src="https://github.com/cypress-io/cypress/assets/154560623/4f710167-b780-444d-b649-18f3ff284263">
<img width="610" alt="unsupported_name" src="https://github.com/cypress-io/cypress/assets/154560623/94979c3a-cd60-4f37-b166-aeea80911602">
<img width="606" alt="unsupported_name_channel" src="https://github.com/cypress-io/cypress/assets/154560623/7540ea07-8108-4c2d-8bdc-8cb93e20d9d6">


### PR Tasks


- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
